main.cpp

```diff
@@ -0,0 +1,181 @@
+#include <iostream>
+#include <vector>
+#include <string>
+#include <sstream>
+#include <unistd.h>
+#include <sys/wait.h>
+#include <cstdlib>
+#include <fstream>
+#include <fcntl.h>
+#include <cstring>
+
+using namespace std;
+
+ofstream history_file("history.txt", ios::app);
+
+vector<string> parse_command(const string& input) {
+    istringstream iss(input);
+    vector<string> args;
+    string token;
+    while (iss >> token) {
+        args.push_back(token);
+    }
+    return args;
+}
+
+string expand_env_variables(const string& input) {
+    string result = input;
+    size_t pos = 0;
+
+    while ((pos = result.find('$', pos)) != string::npos) {
+        size_t end = result.find_first_of(" \t\n", pos + 1);
+        string var_name = result.substr(pos + 1, (end == string::npos ? result.length() : end) - pos - 1);
+        const char* var_value = getenv(var_name.c_str());
+        if (var_value) {
+            result.replace(pos, var_name.length() + 1, var_value);
+        }
+        pos += var_name.length() + (var_value ? strlen(var_value) : 0);
+    }
+    return result;
+}
+
+void executeCommand(vector<string>& args) {
+    vector<char*> c_args(args.size() + 1);
+    for (size_t i = 0; i < args.size(); ++i) {
+        c_args[i] = const_cast<char*>(args[i].c_str());
+    }
+    c_args[args.size()] = nullptr;
+
+
+    if (execvp(c_args[0], c_args.data()) == -1) {
+        perror("execvp");
+        exit(EXIT_FAILURE);
+    }
+}
+
+void handleRedirection(vector<string>& args) {
+    for (size_t i = 0; i < args.size(); ++i) {
+        if (args[i] == ">") {
+            int fd = open(args[i + 1].c_str(), O_WRONLY | O_CREAT | O_TRUNC, 0644);
+            if (fd == -1) {
+                perror("open");
+                exit(EXIT_FAILURE);
+            }
+            dup2(fd, STDOUT_FILENO);
+            close(fd);
+            args.erase(args.begin() + i, args.begin() + i + 2);
```

Comment:
Modifying a vector while running is a bad practice and can lead to unexpected behaviors.